### ISSUE TITLE
ngxs not compatible  with target: es2015

### ISSUE THREAD

**Issue by xmlking** [ngxs/store#773]
change "target": "es5" -->  "target": "es2015", in  tsconfig.json 

```
main.ts:21 TypeError: Class constructor AppState cannot be invoked without 'new'
    at results.push.selectorsToApply.map.a (ngxs-store.js:1386)
    at Array.map (<anonymous>)
    at fn (ngxs-store.js:1386)
    at Store.selectSnapshot (ngxs-store.js:1476)
    at new AppHandler (app.handler.ts:30)
    at core.js:19605
    at _callFactory (core.js:26903)
    at _createProviderInstance (core.js:26846)
    at resolveNgModuleDep (core.js:26805)
    at _callFactory (core.js:26914)

````

**Comment by deebloo** [comment]
Thanks for reporting can you provide the info from the issue template?

**Comment by deebloo** [comment]
@markwhitfeld we should probably add an integration test to make sure it works with es2015

**Comment by xmlking** [comment]
any workaround for this issue?

**Comment by xmlking** [comment]
I cannot try differential loading 😢 
 https://speakerdeck.com/mgechev/2019

**Comment by xmlking** [comment]
It may be my previous environment (Angular CLI version , TypeScript  etc) which caused the above error. But after upgraded to latest Angular CLI 8, I don't face this issue.  we can close this issue.
Thanks 

working sample.... 
https://github.com/xmlking/ng-differential-loading

**Comment by xmlking** [comment]
Note:
when Angular CLI version  <= `0.800.0-beta.10`  we still get this error when    `"target": "es2015",` is set
repo to reproduce: 
https://github.com/xmlking/ngx-starter-kit

**Comment by splincode** [comment]
@xmlking I checked in the latest version and this problem does not reproduce

**Comment by xmlking** [comment]
@splincode how do i get latest version of @ngxs/store ? cont find  latest package  greater then 3.4.3
 in npm?

**Comment by splincode** [comment]
@xmlking

```
npm i @ngxs/store@dev
```

**Comment by xmlking** [comment]
i upgraded to     
```
"@ngxs/devtools-plugin": "^3.4.3-dev.master-a1ac2a6",
"@ngxs/store": "^3.4.3-dev.master-a1ac2a6",
```
please see above https://github.com/ngxs/store/issues/773#issuecomment-486867640 to reproduce 

still getting error 
![image](https://user-images.githubusercontent.com/627951/58207979-a795db00-7c98-11e9-9721-d5fa44868dff.png)

**Comment by splincode** [comment]
@xmlking

How I can reproduce?

**Comment by xmlking** [comment]
Please clone above simplified project and run ‘yarn’ and ‘ng serve’
See error in the console

**Comment by zakton5** [comment]
Having the same issue. NGXS breaks after upgrading to Angular 8. I am on `3.5.0`.

**Comment by splincode** [comment]
@zakton5 can you help me, provide simple repro?

**Comment by zakton5** [comment]
@splincode Use @xmlking 's project but change the `@ngxs/store` dependency to `3.5.0` to be sure that it recreates the same problem.

**Comment by arturovt** [comment]
@xmlking 
I was debugging a little bit. The `AppHandler`'s constructor is invoked before `NgxsModule` is instantiated.

This worked for me:
```ts
// Noop handler for factory function
export function noop(injector: Injector) {
  return () => {
    console.log(injector.get(AppHandler))
  };
}

providers: [
    {
      provide: APP_INITIALIZER,
      useFactory: noop,
      deps: [Injector],
      multi: true,
    },
  ],
```

**Comment by arturovt** [comment]
Digging deeper. The platform is bootstrapped w/o errors, but it causes indescribable behavior.

If we target `es5` then the `installed` property equals `undefined`, it's still not selected from the store.
```ts
const installed = this.store.selectSnapshot(AppState.installed); // undefined
```

In that case `es2015` tells you that something is really wrong, that's why it fails. It fails because targetting `es2015` doesn't force compiler to emit `function` instead of `class`.

So `AppState` looks different in different targets:

```ts
class AppState {} // es2015
var AppState = function() { ... }; // es5
```

The selector tries to access static metadata, if it doesn't exist, the selector returns plain class:
```ts
function getSelectorFn(selector) {
    var metadata = getSelectorMetadata(selector) || getStoreMetadata(selector);
    return (metadata && metadata.selectFromAppState) || selector;
}
```

The `selectFromAppState` property is not defined at this stage, thus it returns `selector` which is `AppState`. Then it tries to invoke the result of `getSelectorFn`, it fails with:
```ts
class AppState {}
AppState(arguments); // es2015 -> fail

var AppState = function() { ... };
AppState(arguments); // es5 -> OK
```

@xmlking I'm confirming it's not an issue in this case as it's related to the DI cycle of Angular itself, see my comment above.

**Comment by arturovt** [comment]
@zakton5 IMO this seems meaningless to say that `3.5.0` doesn't work with Angular 8 w/o providing any reproduction...

**Comment by arturovt** [comment]
@xmlking woudn't you mind closing this issue or there is something else I can do for you? :blush:

**Comment by marcus-sa** [comment]
So is this not gonna be fixed?
ES2015 support is important.

**Comment by arturovt** [comment]
What exactly would you want us to fix? Have you read the discussion above? :slightly_smiling_face:

**Comment by marcus-sa** [comment]
@arturovt I just fixed it by patching the `fesm2015` and `esm2015` bundles.
```
$ yarn add -D patch-package
```

Add this file
> patches/@ngxs+store+3.5.1.patch
```
diff --git a/node_modules/@ngxs/store/esm2015/src/utils/selector-utils.js b/node_modules/@ngxs/store/esm2015/src/utils/selector-utils.js
index 0926560..d1a4e7e 100644
--- a/node_modules/@ngxs/store/esm2015/src/utils/selector-utils.js
+++ b/node_modules/@ngxs/store/esm2015/src/utils/selector-utils.js
@@ -96,7 +96,7 @@ export function createSelector(selectors, originalFn, creationMetadata) {
          * @param {?} argFn
          * @return {?}
          */
-        argFn => argFn(state))));
+        argFn => 'prototype' in argFn ? new argFn(state) : argFn(state))));
         // if the lambda tries to access a something on the
         // state that doesn't exist, it will throw a TypeError.
         // since this is quite usual behaviour, we simply return undefined if so.
diff --git a/node_modules/@ngxs/store/fesm2015/ngxs-store.js b/node_modules/@ngxs/store/fesm2015/ngxs-store.js
index dbd53f6..8965d97 100644
--- a/node_modules/@ngxs/store/fesm2015/ngxs-store.js
+++ b/node_modules/@ngxs/store/fesm2015/ngxs-store.js
@@ -2163,7 +2163,7 @@ function createSelector(selectors, originalFn, creationMetadata) {
          * @param {?} argFn
          * @return {?}
          */
-        argFn => argFn(state))));
+        argFn => 'prototype' in argFn ? new argFn(state) : argFn(state))));
         // if the lambda tries to access a something on the
         // state that doesn't exist, it will throw a TypeError.
         // since this is quite usual behaviour, we simply return undefined if so.

```

and add `patch-package` to your `postinstall` scripts

```
{
	// ...
	"scripts": {
		// ...
		"postinstall": "patch-package"
	}
}
```

**Comment by arturovt** [comment]
Sorry, I do not know what is `patch-package` and how is related. If you've got any problems then why wouldn't you want to open an issue with provided reproduction?

**Comment by marcus-sa** [comment]
Nvm, `@ngxs/store` is completely incompatible with es2015 builds (this has got nothing to do with Angular), since you don't differ between methods and classes when calling them, so differential loading is completely useless.

**Comment by arturovt** [comment]
> Couldn't get it to work anyway.
> NGXS is completely incompatible with es2015 builds (this has got nothing to do with Angular), since you don't differ between methods and classes when calling them, so differential loading is completely useless.
> Basically you can only use NGXS with the Angular Webpack Plugin.
> Guess I'll have to migrate to Akita instead.
> You should probably mention this in the docs instead of just closing the issue.

**If you've got any problems then why wouldn't you want to open an issue with provided reproduction?**

**Comment by arturovt** [comment]
@marcus-sa 

I noticed that all your comments have nothing to do with what I ask. Who are you talking to then?

**Comment by markwhitfeld** [comment]
I think that this is a hard problem to solve and it feels like @arturovt and @marcus-sa are talking a bit cross purposes. @marcus-sa do you think you would be able provide a reproduction? Potentially even a PR to fix the issue?
PS. @marcus-sa You are coming across as quite demanding and entitled, I'm sure you don't mean to. We are all contributors that do this in our own time and do not get paid for this work.

**Comment by marcus-sa** [comment]
@markwhitfeld I didn't mean to be rude, so I'm sorry about that.
The closest I can give to a reproduction would be if you created any Angular project and use the es2015 target while selecting some state.
I used Webpack before where the workaround was simply to set the target to es5.
I've just migrated everything to Bazel, where differential loading is done using Rollup & Babel.
I'm not familiar with the NGXS code, but from what I've taken a look at it requires some refactoring, because you haven't taken into consideration that when invoking a function, it'll most most of the time be a class.
This is pretty much "illegal" in strict mode and in any other transpiler.
E.g Babel enforces the same behavior using `_classCallCheck`
